### ISSUE TITLE
MPS: allocator-state-dependent silently wrong matmul results (10-30% of scale) on 2.7-2.13.0; masked (not root-fixed) by #187441

### ISSUE THREAD

**Issue by haishuo** [pytorch/pytorch#193487]
## 🐛 Describe the bug

*Up-front note: this is a complete drive-by report — every measurement we
have is included below, and the reporting account is unlikely to respond to
follow-ups. We're filing it anyway because the failure mode is silent wrong
numerics, which seemed too severe to leave unreported.*

On Apple Silicon (M2 Max, macOS Darwin 25.5.0), fp32 matmuls whose LHS is a
transpose view can return results wrong by 10–30% of the result's scale,
**depending on MPS caching-allocator state** after prior large allocations.
Wrong results are deterministic within a process but vary with allocation
history; the identical tensors compute correctly in a fresh process (we
verified by saving the corrupting fp32 inputs and replaying them — clean on
every version). Buffer contents are intact (`t.to("mps").cpu()` round-trips
bitwise), so a kernel is misreading certain buffer placements rather than
memory being corrupted.

Observed impact (statistical computing): in an iterative batched Newton
solver (batch 20 × 50k rows × 25 cols, ~10 iterations of batched matmuls +
`cholesky_ex`), one batch element's Gram/Hessian assembles wrong by ~13% of
scale — min eigenvalue −167 on a matrix provably PSD+ridge in exact
arithmetic — driving that element to a silently wrong "converged" optimum
(max|coef| 49 vs 15.6 fp64/fp32-CPU ground truth). Finite, plausible, and
invisible to any NaN/inf guard.

## Reproduction (self-contained; fires on <= 2.12.x)

Verified to print identical corrupted values on **2.7.1, 2.9.0, 2.11.0,
2.12.0, 2.12.1**:

```python
import numpy as np, torch

rng = np.random.default_rng(0)
n, p = 50_000, 25
A  = torch.tensor(rng.standard_normal((n, p)), dtype=torch.float32)
w  = torch.tensor(rng.random(n), dtype=torch.float32)
Aw = A * w.unsqueeze(-1)
ref = Aw.double().T @ A.double()

def err():
    return float(((Aw.to("mps").T @ A.to("mps")).cpu().double() - ref).abs().max())

print("fresh    :", err())      # ~9e-5  (correct)
big = torch.tensor(rng.standard_normal((200_000, p)), dtype=torch.float32)
wb  = torch.tensor(rng.random(200_000), dtype=torch.float32)
((big * wb.unsqueeze(-1)).to("mps").T @ big.to("mps")).cpu()   # trigger
print("triggered:", err())      # 7–28 on scale ~90 (10–30%), per-process deterministic
torch.mps.empty_cache()
print("cleared  :", err())      # ~9e-5 again
```

Additional facts, all measured on the same machine:

- Unweighted `A.T @ A` corrupts identically (err 11–23) — weights are not
  required. Small n (5k) after the trigger: clean.
- `torch.mps.synchronize()` does NOT restore correctness;
  `torch.mps.empty_cache()` does. A `.contiguous()` LHS avoids it in this
  minimal form (in the larger solver the error then surfaces in a different
  strided op instead — materializing one operand is not a general fix).
- An operand freshly **produced on-device** is clean even against a
  CPU-copied RHS; the corrupting operand in the minimal repro is a
  CPU→MPS-copied tensor whose buffer placement follows a large prior
  allocation.
- On **2.13.0 (current stable)** this minimal trigger happens to go quiet
  (allocation paths changed in 2.13), but the bug is still present: the
  batched-solver workload corrupts identically to 2.12.1 (same wrong
  Hessian entries, same wrong optimum). We were unable to distill a
  torch-only trigger for 2.13 before wrapping up; the bisection below
  localizes the behavior change regardless.

## Bisection: masked by #187441, not fixed

Bisecting the solver workload with dated nightly wheels:

- `2.14.0.dev20260622`: **corrupted** (wrong optimum 48.8 vs 15.6 truth)
- `2.14.0.dev20260624`, `...0626`, `...0628`, `2.15.0.dev20260813`: clean
  (16/16 fixture fits land on the fp64 ground truth)

The MPS change in that window is **#187441** ("[MPS] Bucket large
allocations to bound caching-allocator reserved memory during decode",
landed 2026-06-23). That PR is a memory-footprint optimization — it changes
which buffers allocations land in (bucketing + freed-buffer reuse instead of
fresh heaps per growth step). Nothing in it touches a kernel, which suggests
the misreading code path is still present on master and is merely no longer
handed the triggering placements. The 2.13 release branch was cut
2026-06-10, before #187441 — consistent with stable 2.13.0 being affected.

## Suggested follow-ups (no response needed from us)

- Root-causing the misread that #187441 masks seems worthwhile: the bug
  presumably lives in how some matmul path derives array
  descriptors/offsets for buffers in particular heap placements, and any
  future allocator tuning could resurface it.
- 2.13.0 currently returns silently wrong numerics for workloads with this
  allocation profile; a cherry-pick of #187441 (or a real fix) into a
  2.13.x patch release would protect current-stable users.

## Versions

- Apple M2 Max, macOS (Darwin 25.5.0), Python 3.13.13, numpy 2.x
- Affected: torch 2.7.1, 2.9.0, 2.11.0, 2.12.0, 2.12.1, 2.13.0
- Clean (empirically): 2.14.0.dev20260624 and later, 2.15.0.dev20260813

cc @ezyang @gchanan @kadeng @msaroufim @jianyuh @nikitaved @mruberry @walterddr @Lezcano @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

**Comment by Achraf921** [comment]
Data point: this somehow does not reproduce on Apple M4 (24GB) / macOS 26.6.1 with torch 2.12.1. All three probes print identical values (fresh = triggered = cleared, approx 2.5e-2 on scale 90, consistent with normal fp32 accumulation, no allocator-state sensitivity at all). Since the suspect matmul path runs through the MPS framework that ships with the OS, the misread may be OS-version dependent in addition to placement dependent, the original report was Darwin 25.5.

**Comment by MohammadSadeghSalehi** [comment]
Confirming this on a different chip and a newer OS than the original report, with two corrections to the write-up and a sharper localisation.

**Environment:** Apple M1 Pro, macOS 26.6 (Darwin 25.6.0), torch **2.13.0** (current stable), Python 3.12.

### 1. The minimal trigger does *not* go quiet on 2.13.0 here

The report says the minimal repro stops firing on 2.13.0 and only the larger solver workload corrupts. On this machine the minimal repro fires on 2.13.0 exactly as described:

```
fresh                        0/625 elements wrong,  maxrel 2.8e-05
triggered                  579/625 elements wrong,  maxrel 1.86
triggered (repeat)         identical, deterministic
after torch.mps.synchronize()  579/625 still wrong
after torch.mps.empty_cache()    0/625 wrong, exactly restored
```

Deterministic within a process and bit-identical across separate processes. So 2.13.0 is reproducibly affected by the *minimal* trigger, not only by the solver workload.

### 2. A `.contiguous()` LHS does not avoid it

The report notes that materialising the transpose sidesteps the minimal form. It does not here: `Aw.to("mps").T.contiguous() @ A.to("mps")` still gives 536/625 elements wrong. That weakens the "transpose view derives a bad descriptor" reading, since the corruption survives materialising the transpose into a fresh buffer.

### 3. What actually decides it: whether an operand buffer came from a H2D copy

Varying how each operand is produced, all after the same trigger, in one process:

```
copied.T   @ copied     bad = 584/625   maxrel 4.34
ondevice.T @ copied     bad = 571/625   maxrel 4.37
copied.T   @ ondevice   bad = 581/625   maxrel 6.62
ondevice.T @ ondevice   bad =   0/625   maxrel 2.8e-05    <- clean
```

where `ondevice` means the operand was computed on-device (`A_mps * w_mps`) rather than copied with `.to("mps")`. Note `A_mps` itself is a H2D copy in every row, so this is about the *buffer that the matmul reads*, not data lineage.

This refines the report's claim that a device-produced operand is clean against a CPU-copied RHS: here that combination still corrupts (571/625). **Both** operands have to be device-produced for the result to be clean. One H2D-copied operand landing in a post-trigger placement is enough.

### 4. Not a simple offset misread

Comparing the corrupted result against row-shifted reference products rules out the obvious "reads from the wrong offset" explanation:

```
shift    0: max|corrupt - ref|      =    17.37
shift    1: max|corrupt - shifted|  = 25498.52
shift    2:                            25692.19
shift    8:                            25480.31
shift   64:                            25518.22
shift  256:                            25724.59
```

The corrupted result stays near the true product (17 on a scale of ~25000) while being wrong in 584 of 625 entries. It looks like partial or stale accumulation over the reduction, not a wholesale read of the wrong region.

### 5. On the M4 data point above

The earlier comment could not reproduce on M4 / macOS 26.6.1 / torch 2.12.1. We reproduce on M1 Pro / macOS **26.6** / torch 2.13.0, so OS version alone does not explain it; the chip generation looks like the more likely axis, which fits a misread in a kernel selected per-GPU-family.

### Repro used

```python
import numpy as np, torch
rng = np.random.default_rng(0)
n, p = 50_000, 25
A  = torch.tensor(rng.standard_normal((n, p)), dtype=torch.float32)
w  = torch.tensor(rng.random(n), dtype=torch.float32)
Aw = A * w.unsqueeze(-1)
ref = Aw.double().T @ A.double()

def bad(x):
    rel = ((x.cpu().double() - ref).abs() / ref.abs().clamp_min(1e-12))
    return int((rel > 1e-3).sum()), float(rel.max())

print("fresh    :", bad(Aw.to("mps").T @ A.to("mps")))
big = torch.tensor(rng.standard_normal((200_000, p)), dtype=torch.float32)
wb  = torch.tensor(rng.random(200_000), dtype=torch.float32)
((big * wb.unsqueeze(-1)).to("mps").T @ big.to("mps")).cpu()
print("triggered:", bad(Aw.to("mps").T @ A.to("mps")))
torch.mps.empty_cache()
print("cleared  :", bad(Aw.to("mps").T @ A.to("mps")))
```

Happy to run further probes on this machine, since it reproduces deterministically and the original reporter noted they would not be following up.

**Comment by MohammadSadeghSalehi** [comment]
Followed this up on a source build of `main` (a6cfa80) to test the "still present on master, merely no longer handed the triggering placements" hypothesis. On this machine it does not hold up.

**1. Not reachable on main via allocation state.** Sweeping shapes and trigger sizes, comparing MPS fp32 against a CPU fp64 reference before and after a large allocation:

```
       n    p   trig_n       fresh   triggered    ratio
   50000   25   100000    2.53e-04    2.53e-04      1.0
   50000   25   800000    2.53e-04    2.53e-04      1.0
   50000   64   200000    4.42e-03    4.42e-03      1.0
  100000   25   400000    9.37e-04    9.37e-04      1.0
  200000   16   800000    1.38e-04    1.38e-04      1.0
   ... 20 configurations, every ratio 1.0
```

The original minimal repro is also completely flat on main: fresh, triggered, after `synchronize()`, and after `empty_cache()` all give identical values with 0/625 elements wrong.

**2. Restoring pre-#187441 allocation sizes does not resurface it.** I patched `get_allocation_size` in `MPSAllocator.mm` with a local env switch that skips the large-allocation bucketing and returns the plain aligned size, then ran the repro both ways in the same build:

```
bucketing ON  : 0/625 wrong at every probe
bucketing OFF : 0/625 wrong at every probe
```

The switch is doing something measurable, so this is not a no-op: allocating twelve slightly-growing large buffers gives `current_allocated` 140.75 MiB with bucketing on and 139.32 MiB with it off, which is the size-rounding difference.

**Caveat on what that does and does not show.** My switch reverts only the size rounding. #187441 also changed buffer reuse, freed-buffer reuse instead of fresh heaps per growth step, and I did not revert that half, so this is evidence against the masking hypothesis rather than proof. It is entirely possible the reuse change is what moves the placements.

**Where that leaves things.** For reference, I did reproduce the original bug on 2.13.0 on this same machine, so the hardware and OS can exhibit it: M1 Pro, macOS 26.6 (25G72), and there the minimal repro fires with 579/625 elements wrong at up to 186% relative error, contrary to the report's note that 2.13 goes quiet.

So current stable is affected and main appears genuinely not to be, at least by this route. That makes the reporter's second suggestion the actionable one: a cherry-pick of #187441 into a 2.13.x patch release would protect current-stable users, whether or not the underlying misread is ever root-caused.

Happy to run further probes since this reproduces deterministically on 2.13.0 here, including reverting the reuse half of #187441 if someone who knows the allocator can point me at the right thing to disable.

**Comment by brianleake** [comment]
Independent corroboration with a distinct real-world workload, matching this issue's signature exactly.

**Workload**: Video-Depth-Anything (ViT-S) inference inside iw3/nunif's 2D→3D pipeline, torch 2.12.1, M4 / macOS 26.6. The model's 32-frame window forward returns a **completely constant depth map** (std 0.0) instead of depth, silently.

- Allocator-state-dependent: bitwise-identical input tensors (verified by md5) through the same forward give correct output or the constant, depending only on the pattern of prior allocations — frames staged to MPS in batches of 2 trigger it; batches of 1, 3, 4, 5 do not.
- `torch.mps.synchronize()` does not help; cloning or CPU-round-tripping the input does not help; live GPU weights verified bitwise against the checkpoint (no corruption).
- `torch.mps.empty_cache()` immediately before the forward restores the correct result exactly, and subsequent forwards stay healthy — same masking behaviour described here.
- CPU is clean for the identical call pattern.

Downstream report with the full elimination chain: https://github.com/nagadomi/nunif/issues/735. Happy to run the repro against a nightly if another datapoint helps.

**Comment by brianleake** [comment]
Fix verification for the workload in my earlier comment: on `torch 2.15.0.dev20260913` (macOS 26.6 / M4), the same unpatched code and exact trigger pattern (frames staged in batches of 2, plus the larger vitl model that our `empty_cache()` workaround could NOT rescue on 2.12.1) runs **clean** — 0 corrupted frames across all cases, and the output is **bit-identical** to healthy 2.12.1 runs (std 0.231142 in both), so the fix restores exact correctness rather than merely plausible numerics. One additional 2.12.1 datapoint from our case: the `empty_cache()` masking is allocation-size-dependent — it rescued the vits/vitb models but not vitl, where corruption fired at a site beyond the guarded call.

**Comment by MohammadSadeghSalehi** [comment]
Correcting my earlier comment. The cherry-pick candidate I named, #187441, is not what makes `main` clean. The fix is #183535, which routes 2-D `mm`/`addmm`/`linear` with K > 32768 and both output dims >= 16 to PyTorch's own Metal kernels on Apple7/8, and it shipped in 2.14.0. The root cause is the MPSGraph over-read past the end of the reduction dimension described in #177116 and #195163: the last K tile is not clipped, so whatever sits after the operands is summed in. That explains everything observed here, including why M4 does not reproduce, why the error depends on allocator state, and why `empty_cache()` masks it.

**Route, not allocator.** Same machine as before (M1 Pro, macOS 26.6.2), minimal repro from the top of this issue, plus the same product through `bmm` and `einsum`:

| build | mm fresh | mm triggered | bmm triggered | einsum triggered |
|---|---|---|---|---|
| 2.13.0 | 3.5e-02 | 8.2e+00 | 8.2e+00 | 8.2e+00 |
| 2.13.0, `PYTORCH_MPS_PREFER_METAL=1` | 4.4e-01 | 4.4e-01 | 8.2e+00 | 8.2e+00 |
| 2.14.0 | 4.4e-01 | 4.4e-01 | 3.5e-02 | 3.5e-02 |

Absolute max error against fp64, scale 2.5e+04. On 2.13.0 the env var changes only the route and cures `mm` while the batched paths, under identical allocator state, still fail. On 2.14.0 `mm` is immune with and without the trigger. The 4.4e-01 on the Metal route is ordinary fp32 accumulation, within a factor of two of CPU fp32 on the same operands.

**The batched paths are still wrong on 2.14.0 stable.** The `bmm` row above happens to come out clean because my trigger allocates through the 2-D route, which no longer touches MPSGraph. With the reproducer from #195012 on this machine:

```
torch 2.14.0
  2-D  @  fresh                      max error       0.0029   ok
  2-D  @  after dirtying memory      max error       0.0029   ok
  bmm  after dirtying memory         max error 32000001.5972   <-- WRONG
  einsum bik,bkj->bij (same path)    max error 32000001.5972   <-- WRONG
```

So `bmm`, `baddbmm`, `addbmm`, 3-D `matmul` and `einsum` still go through MPSGraph and #195012 is the pending fix. Worth keeping in mind for the workloads reported above: the ViT run being clean on the nightly does not yet cover attention through `bmm` at long K.

**Practical summary for M1/M2 users.** Upgrade to 2.14.0 for the 2-D paths, or on 2.13.x set `PYTORCH_MPS_PREFER_METAL=1`. Batched matmul at K >= 32767 with both output dims >= 16 is not safe on any release yet. This issue could reasonably be closed as a duplicate of #177116 with the remainder tracked in #195163.
